Add table for `datadog_host`

marcus-crane opened the pull request:
# Tables added
```
+------------------------------------+-------------------------------------------------------------------------------------------------------------------------+
| table                              | description                                                                                                             |
+------------------------------------+-------------------------------------------------------------------------------------------------------------------------+
| datadog_host                       | A host is any piece of infrastructure that runs an instance of the Datadog Agent such as a bare metal instance or a VM. |
+------------------------------------+-------------------------------------------------------------------------------------------------------------------------+
```

Closes: #19

This is my first steampipe contribution so I expect I might have overlooked some things but I've tried to adhere to the style of the existing tables where possible.

There are a few things worth pointing out:

* While I've tried to wire up pagination properly, my workplace doesn't actually have enough hosts to properly test that it actually works (the page limit is 1000+)
* I don't really understand the specifics of hydration when it comes to the absence of something. The [Datadog hosts API](https://docs.datadoghq.com/api/latest/hosts/) will apparently serve results for any host that is live within the last 3 hours by default. Presumably if you were to open Steampipe, close it then open it again in a week, Steampipe would still retain older hosts rather than magically setting `up` to false. If Steampipe removes an entry from the table upon trying to rehydrate it and getting a 404, then this isn't such a big deal but either way, it kind of makes the `up` field redundant if it is always set to `true`
* While I tried to craft some more "Advanced" queries, I'm not that great with SQL so some queries can probably be rewritten to be less complex
* I haven't mapped `host_name`, which is available in the hosts API, as it appears to contain the exact same data as `name` based on the description.

Anyway, I didn't worry too much about the edge cases since they can be ironed out over time I'm sure.

You can do quite a lot of fun queries by making use of `jsonb` which I've heard about for a few years now but I've never actually seen it in action before now.

<img width="1291" alt="CleanShot 2022-07-23 at 17 49 34@2x" src="https://user-images.githubusercontent.com/14816406/180592462-3778a0d0-cbb5-4220-8062-02c0fa8e32c5.png">

marcus-crane commented:
@LalitTurbot Heya, just wanted to bump this PR since it's been ready for review for some time 😅

cbruno10 commented:
Hey @marcus-crane , sorry for the very long delay on reviewing, we missed a few open PRs recently, but are circling back on them. We'll have a few done in the next few days.

Thanks again for opening this PR!

cbruno10 commented on datadog/table_datadog_host.go:
```diff
@@ -0,0 +1,101 @@
+		opts.WithFilter(hostNameFilter)
+	}
+
+	opts.WithCount(1000)
```
@marcus-crane You had mentioned you weren't able to test this as you don't have > 1000 hosts, so one suggestion to test pagination in smaller environments is to set this value lower temporarily (like to 1 or 2). Then you can put in debug logging statements in your code in places you expect to hit if you're iterating through pages.

Then if you run your query, e.g., `select name from datadog_host`, as long as you have more than the temporary limit value you set, you should be able to trace pagination through your debug statements. Hope that helps, please let me know if you need anymore help testing this though!

cbruno10 replied:
Can you please add code that will reduce this `WithCount` if the passed in query `limit` is lower than the default of 1000 we set here? Please see https://github.com/turbot/steampipe-plugin-mongodbatlas/blob/main/mongodbatlas/table_mongodbatlas_org_event.go#L205-L208 for an example.

cbruno10 commented on datadog/table_datadog_host.go:
```diff
@@ -0,0 +1,101 @@
+		Name:        "datadog_host",
+		Description: "A host is any piece of infrastructure that runs an instance of the Datadog Agent such as a bare metal instance or a VM.",
+		// There is no Datadog hosts endpoint for getting a single host so instead we use a filter to perform an exact match
+		Get: &plugin.GetConfig{
```
Since we already have the `name` column handled as an optional qual in the `ListConfig`, I think we can remove the `GetConfig` for now.

cbruno10 commented on datadog/table_datadog_host.go:
```diff
@@ -0,0 +1,101 @@
+	}
+
+	opts.WithCount(1000)
+	opts.WithIncludeHostsMetadata(true)
```
Is there any reason that a user would want to exclude this data (and the one below as well), e.g., sensitive data, increases response data payload significantly, redundant info? I'm not advocating for its removal, just wondering what type of data is included when set to `true` as Datadog has made it optional.

cbruno10 commented on datadog/table_datadog_host.go:
```diff
@@ -0,0 +1,101 @@
+	for paging {
+		resp, _, err := apiClient.HostsApi.ListHosts(ctx, opts)
+		if err != nil {
+			plugin.Logger(ctx).Error("datadog_monitor.listMonitors", "query_error", err)
```
```suggestion
			plugin.Logger(ctx).Error("datadog_monitor.listHosts", "query_error", err)
```

cbruno10 commented on datadog/table_datadog_host.go:
```diff
@@ -0,0 +1,101 @@
+			}
+		}
+
+		// Break loop if using a filter, which is guaranteed to be an exact match
```
Do we need this code? If we pass in `name`, will the Datadog API give us back more results outside of that scope?

cbruno10 commented on datadog/table_datadog_host.go:
```diff
@@ -0,0 +1,101 @@
+			return nil, nil
+		}
+		// Break loop if host list is less than the maximum page size of 1000 items
+		if count > 1000 {
```
Does this API's response metadata contain information like the user's API does? In the `datadog_user` table, I can see we use functions like `resp.Meta.Page.GetTotalCount()`, which seems a bit more reliable than checking a hardcoded value.

cbruno10 commented on docs/tables/datadog_host.md:
````diff
@@ -0,0 +1,64 @@
+  tags;
+```
+
+
````
```suggestion
```

cbruno10 commented on datadog/table_datadog_host.go:
```diff
@@ -0,0 +1,101 @@
+			// Top columns
+			{Name: "name", Type: proto.ColumnType_STRING, Description: "Name of the host."},
+			{Name: "id", Type: proto.ColumnType_STRING, Description: "ID of the host."},
+			{Name: "up", Type: proto.ColumnType_BOOL, Description: "Whether the expected metrics for the host are being received or not."},
```
@marcus-crane I have seen your explanation in the PR body and read the API docs at https://docs.datadoghq.com/api/latest/hosts/, but admittedly I still don't quite understand when `up` would be `false`. Do you have any other information or docs on how this field is calculated?

cbruno10 commented on datadog/table_datadog_host.go:
```diff
@@ -0,0 +1,101 @@
+		Columns: []*plugin.Column{
+			// Top columns
+			{Name: "name", Type: proto.ColumnType_STRING, Description: "Name of the host."},
+			{Name: "id", Type: proto.ColumnType_STRING, Description: "ID of the host."},
```
Does the API support filtering on the host ID at all?

misraved replied:
@cbruno10 The API doesn't support filtering on the host ID. As per the docs - `This endpoint allows searching for hosts by name, alias, or tag`

cbruno10 commented on datadog/table_datadog_host.go:
```diff
@@ -0,0 +1,101 @@
+		List: &plugin.ListConfig{
+			Hydrate: listHosts,
+			KeyColumns: plugin.KeyColumnSlice{
+				{Name: "name", Require: plugin.Optional},
```
Can there be multiple hosts with the same name in a single account (or whatever the scope of a host is), or are they guaranteed to be unique?

misraved replied:
@cbruno10 The hostname parameter in my opinion is guaranteed to be unique. For example, when I install a datadog agent in an AWS instance, the host name is equivalent to the instance id. I don't seem to find any document which guarantees the uniqueness of a host.

cbruno10 commented on docs/tables/datadog_host.md:
```diff
@@ -0,0 +1,64 @@
+
+A host is any piece of infrastructure that the Datadog Agent is installed on such as a bare-metal server or a cloud virtual machine, regardless of provider.
+
+## Examples
```
Great examples!

cbruno10 reviewed (changes requested):
@marcus-crane Thanks again for opening this PR and for including such helpful query examples in the doc!

Please see suggestions and questions. If you have any questions, please let us know!

marcus-crane commented:
Ah, hey! Sorry, I hadn't gotten back to addressing your feedback but I hadn't forgotten 😅 It's been a while so I've forgotten some details but I'm happy to open another PR with improvements down the line 😀
